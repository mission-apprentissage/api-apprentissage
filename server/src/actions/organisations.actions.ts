import { notFound } from "@hapi/boom";
import type { ObjectId } from "mongodb";

import { getDbCollection } from "@/services/mongodb/mongodbService.js";

export async function deleteOrganisation(organisationIdToDelete: ObjectId) {
  const organisation = await getDbCollection("organisations").findOne(
    { _id: organisationIdToDelete },
    { projection: { nom: 1 } }
  );
  if (!organisation?.nom) throw notFound();
  await removeOrganisationFromUsers(organisation.nom);
  await getDbCollection("organisations").deleteOne({ _id: organisationIdToDelete });
}

function removeOrganisationFromUsers(organisationName: string) {
  return getDbCollection("users").updateMany(
    { organisation: organisationName },
    { $set: { organisation: null, updated_at: new Date() } }
  );
}
