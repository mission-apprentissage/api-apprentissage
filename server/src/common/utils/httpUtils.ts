import axios, { AxiosInstance, AxiosRequestConfig, CreateAxiosDefaults } from "axios";
import http from "http";
import https from "https";
import iconv from "iconv-lite";
import { compose } from "oleoduc";

import logger from "@/common/logger";

// https://github.com/axios/axios/issues/3845#issuecomment-1040819908
class BufferedHttpAgent extends http.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);

    this.highWaterMark = highWaterMark;
  }
  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}
class BufferedHttpsAgent extends https.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);
    //see https://github.com/nodejs/node/issues/39092
    this.highWaterMark = highWaterMark;
  }

  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}

type FetchOptions = AxiosRequestConfig & {
  highWaterMark: number;
  raw: boolean;
  encoding: string;
};
async function _fetch(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { method = "GET", data, highWaterMark, ...rest } = options;
  logger.debug(`${method} ${url}...`);

  return client.request({
    url,
    method,
    httpAgent: new BufferedHttpAgent({ highWaterMark }),
    httpsAgent: new BufferedHttpsAgent({ highWaterMark }),
    ...(data ? { data } : {}),
    ...rest,
  });
}

async function fetchStream(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, encoding = "UTF-8", ...rest } = options;
  const response = await _fetch(url, { responseType: "stream", ...rest }, client);

  const stream = compose(response.data, iconv.decodeStream(encoding));

  if (raw) {
    return {
      headers: response.headers,
      stream,
    };
  } else {
    return stream;
  }
}

async function fetchData(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, ...rest } = options;
  const response = await _fetch(url, { responseType: "json", ...rest }, client);
  const data = response.data;

  if (raw) {
    return {
      headers: response.headers,
      data,
    };
  } else {
    return data;
  }
}

async function fetchJson(url: string, options: Partial<FetchOptions> = {}) {
  const response = await _fetch(url, { ...options, responseType: "json" });
  return response.data;
}

function addCsvHeaders(filename: string, encoding: string, res: any) {
  res.setHeader("Content-disposition", `attachment; filename=${filename}`);
  res.setHeader("Content-Type", `text/csv; charset=${encoding}`);
}

const getHttpClient = (options: CreateAxiosDefaults<any> = {}) =>
  axios.create({
    timeout: 15000,
    httpAgent: new http.Agent({ keepAlive: true }),
    httpsAgent: new https.Agent({ keepAlive: true }),
    ...options,
  });

export { addCsvHeaders, fetchJson, fetchStream, fetchData, getHttpClient };
