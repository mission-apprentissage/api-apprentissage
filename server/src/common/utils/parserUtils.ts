import { Options, parse } from "csv-parse";
import { isEmpty, pickBy } from "lodash-es";

export const DEFAULT_DELIMITER = ";";

export function parseCsv(options: Options = {}) {
  return parse({
    trim: true,
    delimiter: DEFAULT_DELIMITER,
    columns: true,
    bom: true,
    relax_column_count: true,
    on_record: (record) => {
      return pickBy(record, (v) => {
        return !isEmpty(v) && v.trim().length;
      });
    },
    ...options,
  });
}
