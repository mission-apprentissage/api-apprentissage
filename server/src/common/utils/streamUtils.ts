import { Transform } from "node:stream";

type Options = {
  size: number;
};

export function createBatchTransformStream(opts: Options): Transform {
  let currentBatch: unknown[] = [];

  return new Transform({
    objectMode: true,
    transform(chunk, _encoding, callback) {
      currentBatch.push(chunk);

      if (currentBatch.length >= opts.size) {
        this.push(currentBatch);
        currentBatch = [];
      }

      callback();
    },
    flush(callback) {
      if (currentBatch.length > 0) {
        this.push(currentBatch);
      }
      callback();
    },
  });
}
