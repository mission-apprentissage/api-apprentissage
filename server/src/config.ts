import env from "env-var";

const config = {
  productName: env.get("PUBLIC_PRODUCT_NAME").required().asString(),
  port: env.get("SERVER_PORT").required().asPortNumber(),
  version: env.get("PUBLIC_VERSION").required().asString(),
  env: env.get("ENV").required().asEnum(["local", "recette", "production", "preview", "test"]),
  publicUrl: env.get("PUBLIC_URL").required().asString(),
  email: env.get("EMAIL").required().asString(),
  email_from: "Mission Apprentissage",
  mongodb: {
    uri: env.get("MONGODB_URI").required().asString(),
  },
  log: {
    type: env.get("LOG_TYPE").required().asString(),
    level: env.get("LOG_LEVEL").required().asString(),
  },
  session: {
    secret: env.get("SESSION_SECRET").required().asString(),
    cookieName: "api_session",
    cookie: {
      maxAge: 30 * 24 * 3600000,
      httpOnly: true,
      sameSite: "lax" as const,
      path: "/",
      secure: env.get("SESSION_COOKIE_SECURE").default("true").asBool(),
    },
  },
  api_key: {
    // 3 mois
    expiresIn: 90 * 24 * 60 * 60 * 1000,
  },
  auth: {
    user: {
      jwtSecret: env.get("AUTH_USER_JWT_SECRET").required().asString(),
      // 1 mois: meme valeur que le cookie de session
      expiresIn: 30 * 24 * 3600,
    },
    resetPasswordToken: {
      jwtSecret: env.get("AUTH_PASSWORD_JWT_SECRET").required().asString(),
      expiresIn: "1h",
    },
    hashRounds: env.get("AUTH_HASH_ROUNDS").required().asIntPositive(),
  },
  smtp: {
    host: env.get("SMTP_HOST").required().asString(),
    port: env.get("SMTP_PORT").required().asString(),
    secure: env.get("SMTP_SECURE").asBool(),
    webhookKey: env.get("SMTP_WEBHOOK_KEY").required().asString(),
    auth: {
      user: env.get("SMTP_AUTH_USER").required().asString(),
      pass: env.get("SMTP_AUTH_PASS").required().asString(),
    },
  },
  disable_processors: env.get("DISABLE_PROCESSORS").default("false").asBool(),
  api: {
    acce: {
      username: env.get("API_ACCE_USERNAME").required().asString(),
      password: env.get("API_ACCE_PASSWORD").required().asString(),
    },
    referentielOnisep: {
      endpoint: "https://referentiel.apprentissage.onisep.fr/api/v1",
    },
    catalogue: {
      baseurl: "https://catalogue-apprentissage.intercariforef.org",
    },
    catalogueEducatif: {
      baseurl: "https://catalogue.apprentissage.education.gouv.fr",
      username: env.get("API_CATALOGUE_EDUCATIF_USERNAME").required().asString(),
      password: env.get("API_CATALOGUE_EDUCATIF_PASSWORD").required().asString(),
    },
    entreprise: {
      baseurl: "https://entreprise.api.gouv.fr/v3",
      key: env.get("API_ENTREPRISE_KEY").required().asString(),
      defaultRecipient: "13002526500013", // Siret DINUM
      object: "Consolidation des donn√©es",
      context: "MNA",
    },
  },
};

export default config;
