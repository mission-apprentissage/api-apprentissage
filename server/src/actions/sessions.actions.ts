import { FastifyReply, FastifyRequest } from "fastify";
import jwt from "jsonwebtoken";
import { Filter, FindOptions, ObjectId } from "mongodb";
import { ISession } from "shared/models/session.model";

import config from "@/config";
import { getDbCollection } from "@/services/mongodb/mongodbService";
import { authCookieSession } from "@/services/security/authenticationService";

async function createSession(email: string) {
  const now = new Date();

  const session: ISession = {
    _id: new ObjectId(),
    email,
    updated_at: now,
    created_at: now,
    expires_at: new Date(now.getTime() + config.session.cookie.maxAge),
  };

  await getDbCollection("sessions").insertOne(session);

  return session;
}

async function getSession(filter: Filter<ISession>, options?: FindOptions): Promise<ISession | null> {
  return getDbCollection("sessions").findOne(filter, options);
}

async function deleteSession({ email }: { email: string }) {
  await getDbCollection("sessions").deleteMany({ email });
}

function createSessionToken(email: string) {
  return jwt.sign({ email }, config.auth.user.jwtSecret, {
    issuer: config.publicUrl,
    expiresIn: config.session.cookie.maxAge / 1_000,
    subject: email,
  });
}

async function startSession(email: string, res: FastifyReply) {
  const token = createSessionToken(email);
  await createSession(email);
  res.setCookie(config.session.cookieName, token, config.session.cookie);
}

async function stopSession(req: FastifyRequest, res: FastifyReply) {
  const user = await authCookieSession(req);
  if (user) {
    await deleteSession({ email: user.value.email });
  }

  res.clearCookie(config.session.cookieName, config.session.cookie);
}

export { getSession, startSession, stopSession, createSessionToken, createSession };
