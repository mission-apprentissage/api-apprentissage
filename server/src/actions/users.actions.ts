import { ObjectId } from "mongodb";
import { IUser } from "shared/models/user.model";

import { getDbCollection } from "@/services/mongodb/mongodbService";

import config from "../config";
import { generateKey, generateSecretHash } from "../utils/cryptoUtils";
import { createUserTokenSimple } from "../utils/jwtUtils";

export const updateUser = async (email: IUser["email"], data: Partial<IUser>): Promise<void> => {
  await getDbCollection("users").findOneAndUpdate(
    {
      email,
    },
    {
      $set: { ...data, updated_at: new Date() },
    }
  );
};

export const generateApiKey = async (user: IUser) => {
  const generatedKey = generateKey();
  const secretHash = generateSecretHash(generatedKey);

  await getDbCollection("users").findOneAndUpdate(
    {
      email: user.email,
    },
    {
      $set: { updated_at: new Date() },
      $push: {
        api_keys: {
          _id: new ObjectId(),
          name: null,
          key: secretHash,
          last_used_at: null,
          expires_at: new Date(Date.now() + config.api_key.expiresIn),
        },
      },
    }
  );

  const token = createUserTokenSimple({
    payload: { _id: user._id, api_key: generatedKey },
    expiresIn: config.api_key.expiresIn / 1_000,
  });

  return token;
};
