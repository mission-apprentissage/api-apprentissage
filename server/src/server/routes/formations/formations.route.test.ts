import { parseApiAlternanceToken, zFormation } from "api-alternance-sdk";
import nock, { cleanAll, disableNetConnect, enableNetConnect } from "nock";
import {
  generateFormationInternalFixture,
  generateOrganisationFixture,
  generateUserFixture,
  sourceCommuneFixtures,
} from "shared/models/fixtures/index";
import type { IUser } from "shared/models/user.model";
import { beforeAll, beforeEach, describe, expect, it } from "vitest";
import { useMongo } from "@tests/mongo.test.utils.js";

import { generateApiKey } from "@/actions/users.actions.js";
import config from "@/config.js";
import type { Server } from "@/server/server.js";
import createServer from "@/server/server.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";

useMongo();

beforeEach(() => {
  disableNetConnect();

  return () => {
    cleanAll();
    enableNetConnect();
  };
});

let app: Server;

beforeAll(async () => {
  app = await createServer();
  await app.ready();

  return () => app.close();
}, 15_000);

const organisations = {
  jobWrite: generateOrganisationFixture({
    nom: "Org Job Write",
    slug: "org-job-write",
    habilitations: ["jobs:write"],
  }),
  applicationWrite: generateOrganisationFixture({
    nom: "Org Application Write",
    slug: "org-application-write",
    habilitations: ["applications:write"],
  }),
  read: generateOrganisationFixture({
    nom: "Org Read",
    slug: "org-read",
    habilitations: [],
  }),
  appointmentsWrite: generateOrganisationFixture({
    nom: "Org appointments Write",
    slug: "org-appointments-write",
    habilitations: ["appointments:write"],
  }),
};

const users = {
  basic: generateUserFixture({
    email: "basic@exemple.fr",
    is_admin: false,
    organisation: null,
  }),
  read: generateUserFixture({
    email: "ro@exemple.fr",
    is_admin: false,
    organisation: organisations.read.nom,
  }),
  jobWrite: generateUserFixture({
    email: "job-write@exemple.fr",
    is_admin: false,
    organisation: organisations.jobWrite.nom,
  }),
  applicationWrite: generateUserFixture({
    email: "application-write@exemple.fr",
    is_admin: false,
    organisation: organisations.applicationWrite.nom,
  }),
  appointmentsWrite: generateUserFixture({
    email: "appointments-write@exemple.fr",
    is_admin: false,
    organisation: organisations.appointmentsWrite.nom,
  }),
};

const tokens = {
  basic: "",
  read: "",
  jobWrite: "",
  applicationWrite: "",
  appointmentsWrite: "",
};

describe("GET /formation/v1/search", () => {
  const clichy = {
    longitude: 2.3041,
    latitude: 48.9041,
  };

  const levallois = {
    longitude: 2.2874,
    latitude: 48.8946,
  };

  const formations = [
    generateFormationInternalFixture({
      identifiant: { cle_ministere_educatif: "paris" },
      lieu: {
        geolocalisation: sourceCommuneFixtures["75"][0].centre,
      },
    }),
    generateFormationInternalFixture({
      identifiant: { cle_ministere_educatif: "clichy" },
      lieu: {
        geolocalisation: {
          type: "Point",
          coordinates: [clichy.longitude, clichy.latitude],
        },
      },
    }),
    generateFormationInternalFixture({
      identifiant: { cle_ministere_educatif: "levallois" },
      lieu: {
        geolocalisation: {
          type: "Point",
          coordinates: [levallois.longitude, levallois.latitude],
        },
      },
    }),
    generateFormationInternalFixture({
      identifiant: { cle_ministere_educatif: "archived" },
      lieu: {
        geolocalisation: {
          type: "Point",
          coordinates: [levallois.longitude, levallois.latitude],
        },
      },
      statut: {
        catalogue: "archivé",
      },
    }),
  ];

  beforeEach(async () => {
    await getDbCollection("formation").insertMany(formations);
  });

  it("should returns 401 if api key is not provided", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/api/formation/v1/search",
    });

    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Vous devez fournir une clé d'API valide pour accéder à cette ressource",
    });
  });

  it("should returns 401 if api key is invalid", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/api/formation/v1/search",
      headers: {
        Authorization: `Bearer ${tokens.basic}invalid`,
      },
    });
    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Impossible de déchiffrer la clé d'API",
    });
  });

  it("should perform search correctly", async () => {
    const response = await app.inject({
      method: "GET",
      url: `/api/formation/v1/search?latitude=${clichy.latitude}&longitude=${clichy.longitude}&radius=30`,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });
    expect.soft(response.statusCode).toBe(200);
    const result = response.json();
    expect.soft(result.pagination).toEqual({
      page_count: 1,
      page_size: 100,
      page_index: 0,
    });
    expect.soft(result.data).toHaveLength(3);
    expect
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .soft(result.data.map((r: any) => r.identifiant))
      .toEqual([formations[1].identifiant, formations[2].identifiant, formations[0].identifiant]);
    expect(result.data[0]).toMatchSnapshot();
  });

  it("should support include_archived param", async () => {
    const response = await app.inject({
      method: "GET",
      url: `/api/formation/v1/search?include_archived=true`,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });
    expect.soft(response.statusCode).toBe(200);
    const result = response.json();
    expect.soft(result.pagination).toEqual({
      page_count: 1,
      page_size: 100,
      page_index: 0,
    });
    expect.soft(result.data).toHaveLength(formations.length);
  });

  it("should support include_archived param", async () => {
    const response = await app.inject({
      method: "GET",
      url: `/api/formation/v1/search?include_archived=false`,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });
    expect.soft(response.statusCode).toBe(200);
    const result = response.json();
    expect.soft(result.pagination).toEqual({
      page_count: 1,
      page_size: 100,
      page_index: 0,
    });
    expect.soft(result.data).toHaveLength(3);
  });
});

const nockMatchUserAuthorization = (u: IUser, habilitations: string[]) => {
  let token: string = "";

  return {
    matchHeader: (t: string) => {
      token = t;
      return true;
    },
    expectAuth: async () => {
      return expect
        .soft(
          parseApiAlternanceToken({
            token,
            publicKey: config.api.alternance.public_cert,
          })
        )
        .resolves.toEqual({
          data: {
            email: u.email,
            habilitations: habilitations.reduce((acc, h) => ({ ...acc, [h]: true }), {
              "applications:write": false,
              "appointments:write": false,
              "jobs:write": false,
            }),
            organisation: u.organisation,
          },
          success: true,
        });
    },
  };
};

beforeEach(async () => {
  await getDbCollection("users").insertMany(Object.values(users));
  await getDbCollection("organisations").insertMany(Object.values(organisations));
  tokens.basic = (await generateApiKey("", users.basic)).value;
  tokens.read = (await generateApiKey("", users.read)).value;
  tokens.jobWrite = (await generateApiKey("", users.jobWrite)).value;
  tokens.applicationWrite = (await generateApiKey("", users.applicationWrite)).value;
  tokens.appointmentsWrite = (await generateApiKey("", users.appointmentsWrite)).value;
});

describe("POST /formation/v1/appointment/generate-link", () => {
  const body = { cle_ministere_educatif: "088281P01313885594860007038855948600070-67118#L01" };

  it("should returns 401 if api key is not provided", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/api/formation/v1/appointment/generate-link",
      body: body,
    });

    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Vous devez fournir une clé d'API valide pour accéder à cette ressource",
    });
  });

  it("should returns 401 if api key is invalid", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/api/formation/v1/appointment/generate-link",
      body: body,
      headers: {
        Authorization: `Bearer ${tokens.basic}invalid`,
      },
    });
    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Impossible de déchiffrer la clé d'API",
    });
  });

  it("should returns 403 if user doesn't have organisation", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/api/formation/v1/appointment/generate-link",
      body: body,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });
    expect(response.statusCode).toBe(403);
    expect(response.json()).toEqual({
      statusCode: 403,
      message: "Vous n'êtes pas autorisé à accéder à cette ressource",
      name: "Forbidden",
    });
  });

  it.each<[keyof typeof tokens]>([["read"], ["applicationWrite"], ["jobWrite"]])(
    "should returns 403 if organisation doesn't have habilitation appointment:write (%s)",
    async (name) => {
      const response = await app.inject({
        method: "POST",
        url: "/api/formation/v1/appointment/generate-link",
        body: body,
        headers: {
          Authorization: `Bearer ${tokens[name]}`,
        },
      });
      expect.soft(response.statusCode).toBe(403);
      expect(response.json()).toEqual({
        statusCode: 403,
        message: "Vous n'êtes pas autorisé à accéder à cette ressource",
        name: "Forbidden",
      });
    }
  );

  it("should return result from lba", async () => {
    const data = {
      cfd: "32025001",
      cle_ministere_educatif: "088281P01313885594860007038855948600070-67118#L01",
      code_postal: "93290",
      etablissement_formateur_entreprise_raison_sociale: "AFORP FORMATION",
      etablissement_formateur_siret: "77572845400205",
      form_url:
        "https://labonnealternance.apprentissage.beta.gouv.fr/formation/rdv/088281P01313885594860007038855948600070-67118#L01",
      intitule_long: "ASSISTANCE TECHNIQUE D'INGENIEUR (BTS)",
      lieu_formation_adresse: "64 Avenue de la Plaine de France",
      localite: "Tremblay-en-France",
    };

    const { matchHeader, expectAuth } = nockMatchUserAuthorization(users.appointmentsWrite, ["appointments:write"]);
    nock("https://labonnealternance-recette.apprentissage.beta.gouv.fr/api")
      .post("/v2/appointment", (b) => {
        expect.soft(b).toEqual(body);
        return true;
      })
      .matchHeader("authorization", matchHeader)
      .reply(200, data);

    const response = await app.inject({
      method: "POST",
      headers: { Authorization: `Bearer ${tokens.appointmentsWrite}` },
      url: "/api/formation/v1/appointment/generate-link",
      body,
    });

    await expectAuth();
    expect.soft(response.statusCode).toBe(200);
    const result = response.json();
    expect(result).toEqual(data);
  });
});

describe("GET /formation/v1/:id", () => {
  beforeEach(async () => {
    await getDbCollection("formation").deleteMany({});
  });

  it("should returns 401 if api key is not provided", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/api/formation/v1/23",
    });

    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Vous devez fournir une clé d'API valide pour accéder à cette ressource",
    });
  });

  it("should returns 401 if api key is invalid", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/api/formation/v1/23",
      headers: {
        Authorization: `Bearer ${tokens.basic}invalid`,
      },
    });
    expect(response.statusCode).toBe(401);
    expect(response.json()).toEqual({
      statusCode: 401,
      name: "Unauthorized",
      message: "Impossible de déchiffrer la clé d'API",
    });
  });

  it("should return 404 if formation id does not exist", async () => {
    const invalidId = "paris"; // ou un autre ID qui ne sera pas trouvé
    const response = await app.inject({
      method: "GET",
      url: `/api/formation/v1/${invalidId}`,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });

    expect(response.statusCode).toBe(404);
    expect(response.json()).toEqual({
      statusCode: 404,
      name: "Not Found",
      message: `Aucune formation trouvée pour l'identifiant ${invalidId}`,
    });
  });

  it("should return formation by id", async () => {
    const validId = "cle-me-test";

    const formation = generateFormationInternalFixture({
      identifiant: { cle_ministere_educatif: validId },
      lieu: {
        geolocalisation: {
          type: "Point",
          coordinates: [2.2874, 48.8946],
        },
      },
    });
    await getDbCollection("formation").insertOne(formation);

    const response = await app.inject({
      method: "GET",
      url: `/api/formation/v1/${validId}`,
      headers: {
        Authorization: `Bearer ${tokens.basic}`,
      },
    });

    expect(response.statusCode).toBe(200);

    const result = response.json();

    expect(() => zFormation.parse(result)).not.toThrow();
    expect(result.identifiant.cle_ministere_educatif).toBe(validId);
    expect(result).toMatchSnapshot();
  });
});
