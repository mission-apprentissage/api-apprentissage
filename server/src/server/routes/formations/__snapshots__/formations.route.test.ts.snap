// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /formation/v1/:id > should return formation by id 1`] = `
{
  "certification": {
    "connue": true,
    "valeur": {
      "base_legale": {
        "cfd": {
          "abrogation": "2024-07-01T23:59:59.000+02:00",
          "creation": "2021-09-01T00:00:00.000+02:00",
        },
      },
      "blocs_competences": {
        "rncp": [
          {
            "code": "RNCP36629BC01",
            "intitule": "Analyser et diagnostiquer les besoins du client en matière de gestion de patrimoine privé ou professionnel",
          },
          {
            "code": "RNCP36629BC02",
            "intitule": "Conseiller et commercialiser des montages d’ingénierie patrimoniale privée ou professionnelle",
          },
          {
            "code": "RNCP36629BC03",
            "intitule": "Créer son cabinet, développer et pérenniser son portefeuille client privé ou professionnel",
          },
          {
            "code": "RNCP36629BC04",
            "intitule": "Suivre les réglementations et les procédures d’éthique et de déontologie financières en matière de conseil patrimonial privé ou professionnel",
          },
        ],
      },
      "continuite": {
        "cfd": [
          {
            "code": "16X31336",
            "courant": true,
            "fermeture": "2024-08-31T23:59:59.000+02:00",
            "ouverture": "2021-09-01T00:00:00.000+02:00",
          },
        ],
        "rncp": [
          {
            "actif": true,
            "activation": "2022-07-05T02:00:00.000+02:00",
            "code": "RNCP36629",
            "courant": true,
            "fin_enregistrement": "2024-06-30T23:59:59.000+02:00",
          },
        ],
      },
      "convention_collectives": {
        "rncp": [],
      },
      "domaines": {
        "formacodes": {
          "rncp": [
            {
              "code": "41014",
              "intitule": "41014 : Gestion patrimoine",
            },
            {
              "code": "41007",
              "intitule": "41007 : Gestion actifs",
            },
            {
              "code": "41003",
              "intitule": "41003 : Gestion portefeuille",
            },
            {
              "code": "42133",
              "intitule": "42133 : Gestion immobilière",
            },
          ],
        },
        "nsf": {
          "cfd": {
            "code": "313",
            "intitule": "FINANCES, BANQUE, ASSURANCES",
          },
          "rncp": [
            {
              "code": "313",
              "intitule": "313 : Finances, banque, assurances, immobilier",
            },
          ],
        },
        "rome": {
          "rncp": [
            {
              "code": "C1205",
              "intitule": "Conseil en gestion de patrimoine financier",
            },
          ],
        },
      },
      "identifiant": {
        "cfd": "16X31336",
        "rncp": "RNCP36629",
        "rncp_anterieur_2019": true,
      },
      "intitule": {
        "cfd": {
          "court": "EXPERT EN GESTION DE PATRIMOINE",
          "long": "EXPERT EN GESTION DE PATRIMOINE (IPAC)",
        },
        "niveau": {
          "cfd": {
            "europeen": "7",
            "formation_diplome": "16X",
            "interministeriel": "16X",
            "libelle": "TITRE PROFESSIONNEL HOMOLOGUE OU CERTIFIE",
            "sigle": "TH1-X",
          },
          "rncp": {
            "europeen": "7",
          },
        },
        "rncp": "Expert en gestion de patrimoine",
      },
      "periode_validite": {
        "cfd": {
          "derniere_session": 2024,
          "fermeture": "2024-08-31T23:59:59.000+02:00",
          "ouverture": "2021-09-01T00:00:00.000+02:00",
          "premiere_session": 2022,
        },
        "debut": "2021-09-01T00:00:00.000+02:00",
        "fin": "2024-08-31T23:59:59.000+02:00",
        "rncp": {
          "actif": true,
          "activation": "2022-07-05T02:00:00.000+02:00",
          "debut_parcours": "2022-07-01T00:00:00.000+02:00",
          "fin_enregistrement": "2024-06-30T23:59:59.000+02:00",
        },
      },
      "type": {
        "certificateurs_rncp": [
          {
            "nom": "MINISTERE DU TRAVAIL DU PLEIN EMPLOI ET DE L' INSERTION",
            "siret": "11000007200014",
          },
        ],
        "enregistrement_rncp": "Enregistrement sur demande",
        "gestionnaire_diplome": "DEPPA1",
        "nature": {
          "cfd": {
            "code": "2",
            "libelle": "TITRE PROFESSIONNEL HOMOLOGUE OU CERTIFIE",
          },
        },
        "voie_acces": {
          "rncp": {
            "apprentissage": true,
            "candidature_individuelle": true,
            "contrat_professionnalisation": true,
            "experience": true,
            "formation_continue": true,
            "formation_statut_eleve": true,
          },
        },
      },
    },
  },
  "contact": {
    "email": "mail@mail.fr",
    "telephone": "0123456789",
  },
  "contenu_educatif": {
    "contenu": "Contenu éducatif",
    "objectif": "Apprendre",
  },
  "formateur": {
    "connu": true,
    "organisme": {
      "contacts": [],
      "etablissement": {
        "adresse": {
          "academie": {
            "code": "14",
            "id": "A14",
            "nom": "Rennes",
          },
          "code_postal": "35000",
          "commune": {
            "code_insee": "35238",
            "nom": "Rennes",
          },
          "departement": {
            "code_insee": "35",
            "nom": "Ille-et-Vilaine",
          },
          "label": "34 RUE BAHON-RAULT",
          "region": {
            "code_insee": "53",
            "nom": "Bretagne",
          },
        },
        "creation": "1970-01-05T20:22:01.200+01:00",
        "enseigne": null,
        "fermeture": null,
        "geopoint": {
          "coordinates": [
            -1.6864,
            48.1337,
          ],
          "type": "Point",
        },
        "ouvert": true,
        "siret": "19350030300014",
      },
      "identifiant": {
        "siret": "19350030300014",
        "uai": "0352660B",
      },
      "renseignements_specifiques": {
        "numero_activite": "5335P000935",
        "qualiopi": true,
      },
      "statut": {
        "referentiel": "présent",
      },
      "unite_legale": {
        "actif": true,
        "cessation": null,
        "creation": "1969-12-30T08:03:18.000+01:00",
        "raison_sociale": "LYCEE DES METIERS PIERRE MENDES FRANCE",
        "siren": "193500303",
      },
    },
  },
  "identifiant": {
    "cle_ministere_educatif": "cle-me-test",
  },
  "lieu": {
    "adresse": {
      "academie": {
        "code": "25",
        "id": "A25",
        "nom": "Versailles",
      },
      "code_postal": "92110",
      "commune": {
        "code_insee": "92024",
        "nom": "Clichy",
      },
      "departement": {
        "code_insee": "92",
        "nom": "auts-de-Seine",
      },
      "label": "1T impasse Passoir",
      "region": {
        "code_insee": "11",
        "nom": "Île-de-France",
      },
    },
    "geolocalisation": {
      "coordinates": [
        2.2874,
        48.8946,
      ],
      "type": "Point",
    },
    "precision": 15,
    "siret": null,
    "uai": null,
  },
  "modalite": {
    "annee_cycle": 1,
    "duree_indicative": 1,
    "entierement_a_distance": false,
    "mef_10": null,
  },
  "onisep": {
    "discipline": null,
    "domaine_sousdomaine": null,
    "intitule": null,
    "libelle_poursuite": null,
    "lien_site_onisepfr": null,
    "url": null,
  },
  "responsable": {
    "connu": true,
    "organisme": {
      "contacts": [],
      "etablissement": {
        "adresse": {
          "academie": {
            "code": "14",
            "id": "A14",
            "nom": "Rennes",
          },
          "code_postal": "35000",
          "commune": {
            "code_insee": "35238",
            "nom": "Rennes",
          },
          "departement": {
            "code_insee": "35",
            "nom": "Ille-et-Vilaine",
          },
          "label": "34 RUE BAHON-RAULT",
          "region": {
            "code_insee": "53",
            "nom": "Bretagne",
          },
        },
        "creation": "1970-01-05T20:22:01.200+01:00",
        "enseigne": null,
        "fermeture": null,
        "geopoint": {
          "coordinates": [
            -1.6864,
            48.1337,
          ],
          "type": "Point",
        },
        "ouvert": true,
        "siret": "19350030300014",
      },
      "identifiant": {
        "siret": "19350030300014",
        "uai": "0352660B",
      },
      "renseignements_specifiques": {
        "numero_activite": "5335P000935",
        "qualiopi": true,
      },
      "statut": {
        "referentiel": "présent",
      },
      "unite_legale": {
        "actif": true,
        "cessation": null,
        "creation": "1969-12-30T08:03:18.000+01:00",
        "raison_sociale": "LYCEE DES METIERS PIERRE MENDES FRANCE",
        "siren": "193500303",
      },
    },
  },
  "sessions": [
    {
      "capacite": 10,
      "debut": "2023-09-01T02:00:00.000+02:00",
      "fin": "2024-08-31T02:00:00.000+02:00",
    },
    {
      "capacite": 10,
      "debut": "2024-09-01T02:00:00.000+02:00",
      "fin": "2025-08-31T02:00:00.000+02:00",
    },
    {
      "capacite": 10,
      "debut": "2025-09-01T02:00:00.000+02:00",
      "fin": "2026-08-31T02:00:00.000+02:00",
    },
  ],
  "statut": {
    "catalogue": "publié",
  },
}
`;
